ada_feeding_perception/ada_feeding_perception/table_detection.py

```diff
@@ -0,0 +1,483 @@
+        response: The created response message.
+
+        Returns
+        ----------
+        response: The updated response message based on the request.
+        """
+        self.get_logger().info(f"Incoming service request. data: {request.data}")
+
+        response.success = False
+        response.message = f"Failed to set is_on to {request.data}"
+        with self.is_on_lock:
+            self.is_on = request.data
+            response.success = True
+            response.message = f"Successfully set is_on to {request.data}"
+        return response
+
+    def camera_info_callback(self, msg: CameraInfo) -> None:
+        """
+        Store the latest camera info message.
+
+        Parameters
+        ----------
+        msg: The camera info message.
+        """
+        with self.camera_info_lock:
+            self.camera_info = msg
+
+    def depth_image_callback(self, msg: Union[Image, CompressedImage]) -> None:
+        """
+        Store the latest depth image message.
+
+        Parameters
+        ----------
+        msg: The depth image message.
+        """
+        with self.latest_depth_img_msg_lock:
+            self.latest_depth_img_msg = msg
+
+    def image_callback(self, msg: Union[Image, CompressedImage]) -> None:
+        """
+        Store the latest image message.
+
+        Parameters
+        ----------
+        msg: The image message.
+        """
+        with self.latest_img_msg_lock:
+            self.latest_img_msg = msg
+
+    def run(self) -> None:
+        """
+        Run table detection at a specified fixed rate. This function gets the latest RGB
+        image and depth image messages, fits a plane to the table, and calculates the
+        center and orientation of the table in the camera frame. Then, with the calculated
+        center and orientation, it publishes a PoseStamped message. The PoseStamped
+        messages will continue to get published at a specified fixed rate as table detection
+        continues to run.
+        """
+        # Create a fixed rate to run the loop on
+        # TODO: Replace the rate with correct fixed value or create paremeter for rate
+        rate = self.create_rate(self.update_hz)
+
+        # Run while the context is not shut down
+        while rclpy.ok():
+            # Loop at the specified rate
+            rate.sleep()
+
+            # Check if table detection is on
+            # if not, reiterate
+            with self.is_on_lock:
+                is_on = self.is_on
+            if not is_on:
+                continue
+
+            # Get the latest RGB image message
+            rgb_msg = None
+            with self.latest_img_msg_lock:
+                rgb_msg = self.latest_img_msg
+
+            # Get the latest depth image
+            depth_img_msg = None
+            with self.latest_depth_img_msg_lock:
+                depth_img_msg = self.latest_depth_img_msg
+
+            # Get table depth from camera at the (u, v) coordinate (320, 240)
```

Comment:
This is incorrect, right? Doesn't `fit_table` return a depth array, not a depth at a specific point?